design/coroutines.md

```diff
@@ -0,0 +1,688 @@
+ACTOR template <class T, class F>
+Future<Void> filter(FutureStream<T> input, F pred, PromiseStream<T> output) {
+	loop {
+		try {
+			T nextInput = waitNext(input);
+			if (pred(nextInput))
+				output.send(nextInput);
+		} catch (Error& e) {
+			if (e.code() == error_code_end_of_stream) {
+				break;
+			} else
+				throw;
+		}
+	}
+
+	output.sendError(end_of_stream());
+
+	return Void();
+}
+```
+
+A `FutureStream` can be converted into an `AsyncGenerator` by using a simple helper function:
+
+```c++
+template <class T>
+AsyncGenerator<T> toGenerator(FutureStream<T> stream) {
+    loop {
+        try {
+            co_yield co_await stream;
+        } catch (Error& e) {
+            if (e.code() == error_code_end_of_stream) {
+                co_return;
+            }
+            throw;
+        }
+    }
+}
+```
+
+### Generators and Ranges
+
+`Generator<T>` can be used like an input iterator. This means, that it can also be used with `std::ranges`. Consider
+the following coroutine:
+
+```c++
+// returns base^0, base^1, base^2, ...
+Generator<double> powersOf(double base) {
+    double curr = 1;
+    loop {
+        co_yield curr;
+        curr *= base;
+    }
+}
+```
+
+We can use this now to generate views. For example:
+
+```c++
+for (auto v : generatorRange(powersOf(2)) 
+            | std::ranges::views::filter([](auto v) { return v > 10; })
+            | std::ranges::views::take(10)) {
+    fmt::print("{}\n", v);
+}
+```
+
+The above would print all powers of two between 10 and 2^10.
+
+### Eager vs Lazy Execution
+
+One major difference between async generators and tasks (coroutines returning only one value through `Future`) is the
+execution policy: An async generator will immediately suspend when it is called while a task will immediately start
+execution and needs to be explicitly scheduled.
+
+This is a conscious design decision. Lazy execution makes it much simpler to reason about memory ownership. For example,
+the following is ok:
+
+```c++
+Generator<StringRef> randomStrings(int minLen, int maxLen) {
+    Arena arena;
+    auto buffer = new (arena) uint8_t[maxLen + 1];
+    while (true) {
+        auto sz = deterministicRandom()->randomInt(minLen, maxLen + 1);
+        for (int i = 0; i < sz; ++i) {
+            buffer[i] = deterministicRandom()->randomAlphaNumeric();
+        }
+        co_yield StringRef(buffer, sz);
+    }
+}
+```
+
+The above coroutine returns a stream of random strings. The memory is owned by the coroutine and so it always returns
+a `StringRef` and then reuses the memory in the next iteration. This makes this generator very cheap to use, as it only
+does one allocation in its lifetime. With eager execution, this would be much harder to write (and reason about): the
+coroutine would immediately generate a string and then eagerly compute the next one when the string is retrieved.
+However, in Flow a `co_yield` is guarantee to suspend the coroutine until the value was consumed (this is not generally
```

Comment:
Do you mean use `co_yield` within a flow's `ACTOR` function?

Reply:
No, `flow` provides now an implementation for `co_yield`. Terminology is a bit confusing